Framework/DataHandling/test/LoadSampleShapeTest.h

```diff
@@ -0,0 +1,211 @@
+  }
+
+  void test_output_workspace_has_MeshObject_1WS() {
+    loadMeshObject("InputWS", "InputWS", "cube.stl");
+    clearWorkspaces("InputWS", "InputWS");
+  }
+
+  void test_cube() {
+    auto cube = loadMeshObject("InputWS", "InputWS", "cube.stl");
+    TS_ASSERT(cube->hasValidShape());
+    TS_ASSERT_EQUALS(cube->numberOfVertices(), 8);
+    TS_ASSERT_EQUALS(cube->numberOfTriangles(), 12);
+    TS_ASSERT_DELTA(cube->volume(), 3000, 0.001);
+    clearWorkspaces("InputWS", "InputWS");
+  }
+
+  void test_cylinder() {
+    auto cylinder = loadMeshObject("InputWS", "InputWS", "cylinder.stl");
+    TS_ASSERT(cylinder->hasValidShape());
+    TS_ASSERT_EQUALS(cylinder->numberOfVertices(), 722);
+    TS_ASSERT_EQUALS(cylinder->numberOfTriangles(), 1440);
+    TS_ASSERT_DELTA(cylinder->volume(), 589, 1);
+    clearWorkspaces("InputWS", "InputWS");
+  }
+
+  void test_tube() {
+    auto tube = loadMeshObject("InputWS", "InputWS", "tube.stl");
+    TS_ASSERT(tube->hasValidShape());
+    TS_ASSERT_EQUALS(tube->numberOfVertices(), 1080);
+    TS_ASSERT_EQUALS(tube->numberOfTriangles(), 2160);
+    TS_ASSERT_DELTA(tube->volume(), 7068, 1);
+    clearWorkspaces("InputWS", "InputWS");
+  }
+
+private:
+  // Create workspaces and add them to algorithm properties
+  void prepareWorkspaces(LoadSampleShape &alg, const std::string &inputWS,
+                         const std::string &outputWS) {
+    auto inWs =
+        WorkspaceCreationHelper::create2DWorkspaceWithFullInstrument(10, 4);
+    Mantid::API::AnalysisDataService::Instance().add(inputWS, inWs);
+    alg.setPropertyValue("InputWorkspace", inputWS);
+    alg.setPropertyValue("OutputWorkspace", outputWS);
+  }
+
+  void clearWorkspaces(const std::string &inputWS,
+                       const std::string &outputWS) {
+    TS_ASSERT_THROWS_NOTHING(
+        Mantid::API::AnalysisDataService::Instance().remove(inputWS));
+    if (outputWS != inputWS) {
+      TS_ASSERT_THROWS_NOTHING(
+          Mantid::API::AnalysisDataService::Instance().remove(outputWS));
+    }
+  }
+
+  const MeshObject *loadMeshObject(const std::string &inputWS,
+                                   const std::string &outputWS,
+                                   const std::string filename) {
+    LoadSampleShape testLoadSampleShape;
+    testLoadSampleShape.initialize();
+    testLoadSampleShape.setPropertyValue("Filename", filename);
+    prepareWorkspaces(testLoadSampleShape, inputWS, outputWS);
+    TS_ASSERT_THROWS_NOTHING(testLoadSampleShape.execute());
+    TS_ASSERT(testLoadSampleShape.isExecuted());
+    return getMeshObject(outputWS);
+  }
+
+  const MeshObject *getMeshObject(const std::string &outputWS) {
+    MatrixWorkspace_sptr ws = getWorkspace(outputWS);
+    Sample s;
+    TS_ASSERT_THROWS_NOTHING(s = ws->sample());
+    auto &obj = s.getShape();
+    auto mObj = dynamic_cast<const MeshObject *>(&obj);
+    TSM_ASSERT_DIFFERS("Shape is not a mesh object", mObj, nullptr);
+    return mObj;
```

Comment:
Be careful! think about what you have done here. You have taken a reference and have tried to extended it's lifetime beyond the scope of the function in which it was obtained by taking it's address as a pointer and returning it. You only get away with this because the current internal implementation of `Sample` stores shape as a field and it is therefore tied to the lifetime of `Sample` [see](https://github.com/mantidproject/mantid/blob/master/Framework/API/src/Sample.cpp#L106) and only because the same thing happens in `ExperimentInfo` [see](https://github.com/mantidproject/mantid/blob/master/Framework/API/src/ExperimentInfo.cpp#L604).

This is not good. You need to do whatever you need to with the `MeshObject` within the same scope in which you obtain it.

Talk to me if this is not clear.

Reply:
Would the use of shared pointer to MeshObject fix this?

Reply:
Yes, but I don't think you can get one since I think `getShape` returns a `ref` to the `Object`, and I don't think you should be modifying the API of `Sample` to get around this.

Reply:
It seems to be causing a problem, when I'm no longer using the ADS.

Reply:
Hopefully fixed in the course of removing the ADS.